lib/rubocop/ast/fast_array.rb

```diff
@@ -0,0 +1,51 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module AST
+    # FastArray represents a frozen `Array` with fast lookup
+    # using `include?`.
+    # Like `Set`, the case equality `===` is an alias for `include?`
+    #
+    #     FOO = FastArray[:hello, :world]
+    #     FOO.include?(:hello) # => true, quickly
+    #
+    #     case bar
+    #     when FOO # Note: no splat
+    #       # decided quickly
+    #     # ...
+    class FastArray < ::Array
+      attr_reader :to_set
+
+      def initialize(ary)
+        raise ArgumentError, 'Must be initialized with an array' unless ary.is_a?(Array)
+
+        super
+        freeze
+      end
+
+      def self.[](*values)
+        new(values)
+      end
+
+      def freeze
+        @to_set ||= Set.new(self).freeze
+        super
+      end
+
+      # Return self, not a newly allocated FastArray
+      def to_a
+        self
+      end
+
+      def include?(value)
+        @to_set.include?(value)
+      end
+
+      alias === include?
+    end
+  end
+end
+
+def FastArray(list) # rubocop:disable Naming/MethodName
```

Comment:
I am not too crazy about us defining constants outside of the `RuboCop` namespace. Is it necessary? As far as I can see, we only use `FastArray` _inside_ the `RuboCop::AST` namespace.

Reply:
I still think it's going to be better to define some refinement for use within RuboCop's code base and just use a conversion method like `.to_fa` everywhere.

Reply:
The constant `FastArray` is defined within `RuboCop`, but I also added a global method `FastArray`, but it isn't strictly necessary, I could use `FastArray[ary]`. I can try with refinements too. Is `to_fast_array` ok? I fear `to_fa` would be quite cryptic

Reply:
In my personal opinion, if we could get away with defining the `FastArray` method inside `RuboCop` scope, that would be optimal.

Reply:
> In my personal opinion, if we could get away with defining the `FastArray` method inside `RuboCop` scope, that would be optimal.

We could, but we'll need an explicit `extend FastArray::Method` or similar (not needed when inheriting from a class that already extend it, like `Cop::Base`). Let me try that first.

Reply:
Force pushed without a global function, and on the rubocop side too.